apps/extension/src/offscreen.ts

```diff
@@ -0,0 +1,80 @@
+import type { JsonValue } from '@bufbuild/protobuf';
+import { OffscreenRequest } from '@penumbra-zone/types/src/internal-msg/offscreen-types';
+
+export const isOffscreenRequest = (req: unknown): req is OffscreenRequest =>
+  req != null &&
+  typeof req === 'object' &&
+  'type' in req &&
+  typeof req.type === 'string' &&
+  req.type === 'BUILD_ACTION';
+
+export const offscreenMessageHandler = (
+  req: unknown,
+  _: chrome.runtime.MessageSender,
+  sendResponse: (x: JsonValue) => void,
+) => {
+  if (!isOffscreenRequest(req)) return false;
+
+  buildActionHandler(req, sendResponse);
+  return true;
+};
+
+chrome.runtime.onMessage.addListener(offscreenMessageHandler);
+
+export const buildActionHandler = (
+  jsonReq: OffscreenRequest,
+  responder: (r: JsonValue) => void,
+) => {
+  // Destructure the data object to get individual fields
+  const { transactionPlan, witness, fullViewingKey, length } = jsonReq.request;
+
+  // Spawn web workers
+  const workerPromises: Promise<JsonValue>[] = [];
+  for (let i = 0; i < length; i++) {
+    workerPromises.push(spawnWorker(transactionPlan, witness, fullViewingKey, i));
+  }
+
+  // Wait for promises to resolve and construct response format
+  void Promise.all(workerPromises).then(batchActions => {
+    responder({
+      type: 'BUILD_ACTION',
+      data: batchActions,
+    });
+  });
+};
+
+const spawnWorker = (
+  transactionPlan: JsonValue,
+  witness: JsonValue,
+  fullViewingKey: string,
+  actionId: number,
+): Promise<JsonValue> => {
+  return new Promise((resolve, reject) => {
+    const worker = new Worker(new URL('./wasm-task.ts', import.meta.url));
+
+    const onWorkerMessage = (e: MessageEvent) => {
+      resolve(e.data as JsonValue);
+      worker.removeEventListener('message', onWorkerMessage);
+      worker.removeEventListener('message', onWorkerError);
+      worker.terminate();
+    };
+
+    const onWorkerError = (error: unknown) => {
+      reject(error);
+      worker.removeEventListener('message', onWorkerMessage);
+      worker.removeEventListener('message', onWorkerError);
+      worker.terminate();
+    };
+
+    worker.addEventListener('message', onWorkerMessage);
+    worker.addEventListener('error', onWorkerError);
+
+    // Send data to web worker
+    worker.postMessage({
+      transactionPlan,
+      witness,
+      fullViewingKey,
+      actionId,
+    });
```

Comment:
Please add `satisfies` typing to this. Is it supposed to be `OffscreenRequestPayload`?

Reply:
L incorporated this change